source/client-side-encryption/tests/README.rst

```diff
@@ -2188,3 +2188,142 @@ attached to the virtual machine.
+
+For each test case, follow the process for obtaining the token as outlined in
+the `automatic Azure credentials section <auto-azure_>`_ with the following
+changes:
+
+1. Instead of the standard IMDS TCP endpoint of `169.254.169.254:80`,
+   communicate with the running ``fake_azure`` HTTP server.
+
+2. For each test case, the behavior of the server may be controlled by attaching
+   an additional HTTP header to the sent request: ``X-MongoDB-HTTP-TestParams``.
+
+
+Case 1: Success
+```````````````
+
+Do not set an ``X-MongoDB-HTTP-TestParams`` header.
+
+Upon receiving a response from ``fake_azure``, the driver must decode the
+following information:
+
+1. HTTP status will be ``200 Okay``.
+2. The HTTP body will be a valid JSON string.
+3. The access token will be the string ``"magic-cookie"``.
+4. The expiry duration of the token will be seventy seconds.
+5. The token will have a resource of ``"https://vault.azure.net"``
+
+
+Case 2: Empty JSON
+``````````````````
+
+This case addresses a server returning valid JSON with invalid content.
+
+Set ``X-MongoDB-HTTP-TestParams`` to ``case=empty-json``.
+
+Upon receiving a response:
+
+1. HTTP status will be ``200 Okay``
+2. The HTTP body will be a valid JSON string.
+3. There will be no access token, expiry duration, or resource.
+
+The test case should ensure that this error condition is handled gracefully.
+
+
+Case 3: Bad JSON
+````````````````
+
+This case addresses a server returning malformed JSON.
+
+Set ``X-MongoDB-HTTP-TestParams`` to ``case=bad-json``.
+
+Upon receiving a response:
+
+1. HTTP status will be ``200 Okay``
+2. The response body will contain a malformed JSON string.
+
+The test case should ensure that this error condition is handled gracefully.
+
+
+Case 4: HTTP 404
+````````````````
+
+This case addresses a server returning a "Not Found" response. This is
+documented to occur spuriously within an Azure environment.
+
+Set ``X-MongoDB-HTTP-TestParams`` to ``case=404``.
+
+Upon receiving a response:
+
+1. HTTP status will be ``404 Not Found``.
+2. The response body is unspecified.
+
+The test case should ensure that this error condition is handled gracefully.
+
+
+Case 5: HTTP 500
+````````````````
+
+This case addresses an IMDS server reporting an internal error. This is
+documented to occur spuriously within an Azure environment.
+
+Set ``X-MongoDB-HTTP-TestParams`` to ``case=500``.
+
+Upon receiving a response:
+
+1. HTTP status code will be ``500``.
+2. The response body is unspecified.
+
+The test case should ensure that this error condition is handled gracefully.
+
+
+Case 6: Slow Response
+`````````````````````
+
+This case addresses an IMDS server responding very slowly. Drivers should not
+halt the application waiting on a peer to communicate.
+
+Set ``X-MongoDB-HTTP-TestParams`` to ``case=slow``.
+
+The HTTP response from the ``fake_azure`` server will take at least 1000 seconds
+to complete. The client should report a timeout within a few seconds.
```

Comment:
I suggest adding the timeout in the specification to favor consistency between driver implementations. Since C is [using 3 seconds](https://github.com/mongodb/mongo-c-driver/blob/d217ff83f852f14c9cb7b0e400b4ce983c2edf2e/src/libmongoc/src/mongoc/mcd-azure.c#L182), consider adding:

> 6. Issue `Req` to the Azure IMDS server ``168.254.169.254:80`` **with a timeout of 3 seconds**.

Reply:
can a timeout be 10 seconds? We already have this timeout in other places and based on c# guidelines, it's better to use a single http client for all http requests. Changing it on our side will be a BC

Reply:
I do not feel strongly on any particular timeout duration. Three seconds is what I happened to write into the test, but it can be changed.